### PR TITLE
Quadratic Objective Function to support basic vector operations

### LINKED ISSUE #819: Quadratic Objective Function to support basic vector operations
### Feature request

It would be quite nice if the quadratic objective function supported basic vector operations (vectors in the mathematical sense) such as addition and scalar multiplication. It's quite common to want to combine various quadratic objective functions possibly with a weighting. A proposed API would be:

```csharp
var target1 = new QuadraticObjectiveFunction(quadraticTerms1, linearTerms1);
var target2 = new QuadraticObjectiveFunction(quadraticTerms2, linearTerms2);
double weight1 = 0.4, weight2 = 0.6;

QuadraticObjectiveFunction weightedTarget = (weight1 * target1) + (weight2 * target2);
```

Mathematically, this is trivial because `f=x(Q+P)x+(c+d)x=xQx+xPx+cx+dc=f1+f2` but, syntactically, it would be really nice to write the maths and not have to scale and add individual matrices and vectors etc. If you like it, I'll add it. It's probably a quick job but there are some dictionaries floating around in that function that will need to be handled with care.

p.s. - [This line](https://github.com/accord-net/framework/blob/development/Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs#L351) looks a bit suspicious to me. That only works if Q is symmetric but the class permits you to pass in a non-symmetric matrix. It should either calculate 0.5*(Q+Q')*x or throw an `Exception` if non-symmetric matrices are passed in. I would propose the latter. Assuming the latter, `Q.Dot(input)` would be much faster.

Thanks,
Alex

### PR DESCRIPTION
**Pull request by AlexJCross**
Hi @cesarsouza,

I know I can merge in directly but I'd really appreciate your thoughts on 3 points.

- Throws `NonSymmetricMatrixException` if relative tolerance is > `1e-6`. Does that sound reasonable to you? I didn't want the requirement for the matrix to be symmetric to be too much of a burden but i can still envisage scenarios where this might throw. Any better suggestions 100% welcome. Alternatively, if it detects a non-symmetric matrix, it could always grab the non-symmetric part and continue.
 - Documentation as requested :)  I have tried my best to add XML documentation but I am afraid I couldn't completely fathom out SandCastle. [Please would you confirm I have done this correctly](https://github.com/AlexJCross/framework/commit/2c9517ad0f6debfe8025826caccee9c1fd9bdb49#diff-812c0c8daaf325a8e066ce9bf8c2e30aR155)? Thank you!
 - As discussed (#832), I changed the default variable names to improve readability. However, thinking about it a bit more, I want to check you are happy with this change. If anyone was indexing into the `IDictionary<string, int> Variables`, their code would now break. It's also not consistent with the `NonlinearObjectiveFunction`. So I can: leave it as is; change the Nonlinear as well for consistency; revert completely; or simply change the `ToString()` method aesthetically but not touch the underlying variable names.

Reminder:
```csharp
before: 4x0x0 +2x1x1 +3x2x2 +3x0x1 +1x0x2 +2x1x2
after : 4x₀² +2x₁² +3x₂² +3x₀x₁ +1x₀x₂ +2x₁x₂
```

Happy to make changes as you see fit.

Thanks!
Alex

ps - this closes #819

### CODE REVIEW

File: Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs [hunk 1]
```diff
@@ -127,20 +145,23 @@ namespace Accord.Math.Optimization
     ///   double value = solver.Value;
     ///   double[] solutions = solver.Solution;
     /// </code>
+    /// 
+    /// <para>
+    ///   Sometimes it easiest to compose quadratic functions as linear
```
Comment (cesarsouza) [inline Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs]: Is there a typo here? Sorry I am not a native English speaker, so I thought it was better to ask :-)

File: Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs [hunk 2]
```diff
@@ -174,18 +191,21 @@ public double ConstantTerm
         /// 
         public QuadraticObjectiveFunction(double[,] quadraticTerms, double[] linearTerms, params string[] variables)
         {
-            if (quadraticTerms.GetLength(0) != quadraticTerms.GetLength(1))
+            if (!quadraticTerms.IsSquare())
                 throw new DimensionMismatchException("quadraticTerms", "The matrix must be square.");
 
-            if (quadraticTerms.GetLength(0) != linearTerms.Length)
+            if (!quadraticTerms.IsSymmetric(rtol: 1e-6))
```
Comment (cesarsouza) [inline Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs]: I think this is fine. If for some reason it throws for a matrix that was indeed supposed to be symmetric, the user can always "force" it to be symmetric with [GetSymmetric](http://accord-framework.net/docs/html/M_Accord_Math_Matrix_GetSymmetric__1_1.htm) or manually.

File: Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs [hunk 3]
```diff
@@ -127,20 +145,23 @@ namespace Accord.Math.Optimization
...
     ///   double[] solutions = solver.Solution;
     /// </code>
+    /// 
+    /// <para>
+    ///   Sometimes it easiest to compose quadratic functions as linear
+    ///   combinations of other quadratic functions. The <see cref="QuadraticObjectiveFunction" />
+    ///   supports this by overloading the addition and multiplication operators.
+    /// </para>
+    /// 
+    /// <code source="Unit Tests\Accord.Tests.Math\Optimization\QuadraticObjectiveFunctionTest.cs" region="doc_example" />
```
Comment (cesarsouza) [inline Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs]: It is perfect, thanks for following the style so closely!

File: Unit Tests/Accord.Tests.Math/Optimization/QuadraticObjectiveFunctionTest.cs [hunk 4]
```diff
@@ -144,11 +144,11 @@ public void LinearTest()
         [Test]
         public void HomogeneousTest()
         {
-            double[,] quadraticTerms = 
+            double[,] quadraticTerms =
             {
-                {  1, 3, 1 },
-                {  3, 2, 2 },
-                {  1, 2, 3 },
+                {  8, 3, 1 },
```
[the remark this answers is not in the document]
Comment (AlexJCross) [reply Unit Tests/Accord.Tests.Math/Optimization/QuadraticObjectiveFunctionTest.cs]: Ah I see :D I must admit I spent more than 20 mins trying to remember why this was actually changed (it wasn't really a deliberate change). Anyways, I remember now...

The `ToString()` method only displays the coefficients to the nearest integer (it's always been like that - not a change from me). As that test has odd numbers down the diagonal, the `ToString()` looked funny. I just switched it to even numbers to verify I hadn't broken it. I can revert it if you like but I presume(d) those numbers were arbitrary.

File: Unit Tests/Accord.Tests.Math/Optimization/QuadraticObjectiveFunctionTest.cs [hunk 5]
```diff
@@ -242,8 +241,8 @@ public void FunctionTest()
                     double e = expected();
 
                     Assert.AreEqual(e, a, 1e-10);
-                    Assert.IsFalse(Double.IsNaN(a));
-                    Assert.IsFalse(Double.IsNaN(e));
+                    Assert.IsFalse(double.IsNaN(a));
```
[the remark this answers is not in the document]
Comment (AlexJCross) [reply Unit Tests/Accord.Tests.Math/Optimization/QuadraticObjectiveFunctionTest.cs]: 👍

File: Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs (hunk 1 above)
Comment (AlexJCross) [reply Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs]: There was indeed. Thanks for spotting. I have fixed the typo and pushed. BTW - if you want a native speaker to review your arXiv article I am more than happy to. However, to be completely clear, your English seems absolutely faultless - and I make typos :) - so I'm not sure I see the need!

File: Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs (hunk 2 above)
Comment (AlexJCross) [reply Sources/Accord.Math/Optimization/QuadraticObjectiveFunction.cs]: Cool - my thinking was similar and I had spotted that method - would work well. Another pair of methods that would be useful would be `GetSymmetricPart()` and `GetAntiSymmetricPart` evaluating `0.5(M+M')` and `0.5(M-M')` respectively.

**Comment by AlexJCross** [comment]
@cesarsouza  as always, you're welcome. Thanks for accepting this PR!

**Pull request merged**